### PR TITLE
Add ‘#[serial]’ tests (v2)

### PR DESCRIPTION
**Pull request by shivjm**
(Fixed version of #42626; sorry for the noise.)

This PR adds a `#[serial]` attribute that marks tests that should be executed one-by-one rather than concurrently (per #33519).

I’ve confirmed that the libtest and tidy tests pass. I wasn’t able to run the full suite as a lot of the run_make tests seem to fail on my Windows system, but everything passed up to that point.

Rust (and PR!) newbie here, so please do let me know if I’m doing something poorly or incorrectly!

### CODE REVIEW

**Comment by leonardo-m** [comment]
Perhaps a name like `#[serial_test]` is better?

**Comment by Mark-Simulacrum** [comment]
I think `#[serial]` is fine given that `#[test]` will also be there, but I could see an argument being made that serial is too generic a name. I do think this needs a feature gate though, which I believe hasn't yet been implemented. I could be wrong though.

**Comment by shivjm** [comment]
Ah, I didn’t understand how the feature gates work and the example I followed (#42219) is apparently a WIP. Looking at #42088, I need to gate this by:

1. Adding an entry to the `declare_features` block in feature_gate.rs with a feature name, a version, and an issue number:
 ```rust
(active, serial_tests, "1.19.0", Some(42684))
```
2. Specifying `AttributeGate::Gated` for the `serial` attribute in `BUILTIN_ATTRIBUTES` (same file), which needs a feature name, an explanation, and a function to check its configuration (auto-generated using `cfg_fn`):
 ```rust
("serial", Normal, Gated(Stability::Unstable, "serial_tests", "the `#[serial]` attribute is an experimental feature", cfg_fn!(serial_tests))
```
3. Adding a new document under `unstable-book/src/language-features/` that points to this PR.
4. Adding an entry under `unstable-book/src/SUMMARY.md` that points to that document.

Is this correct? Have I missed anything? I’ll try to get it done pronto, but if not, it will be about a week.

Re: the name of the attribute itself, I can change that if `serial` is too generic.

File: src/libsyntax/test.rs
Comment (bjorn3) [inline src/libsyntax/test.rs]: Please end the line with a comma, so there is less noise when new fields get added.

File: src/libsyntax/test.rs
Comment (bjorn3) [inline src/libsyntax/test.rs]: Same here

File: src/libtest/lib.rs
Comment (bjorn3) [inline src/libtest/lib.rs]: Same here

File: src/libtest/lib.rs
Comment (bjorn3) [inline src/libtest/lib.rs]: Same here

File: src/libtest/lib.rs
Comment (bjorn3) [inline src/libtest/lib.rs]: Please print the result too. Eg
```rust
panic!("result not okay: {:?}", result),
```

File: src/libtest/lib.rs
Comment (bjorn3) [inline src/libtest/lib.rs]: Same printing thing here.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
Few minor things. Overal code looks good to me.

**Comment by Mark-Simulacrum** [comment]
@shivjm I believe this is waiting on some changes from you and a rebase. Let us know if we can help in any way! Feel free to ping me on IRC (simulacrum).

**Comment by shivjm** [comment]
Sorry, been out of town and then busy catching up! I’ve rebased the PR and (I hope) fixed the issues @bjorn3 pointed out. Next up, need to implement the feature gate. I’ll ping you on IRC.

**Comment by shivjm** [comment]
I’ve added the `serial_tests` feature gate. @est31, does this need an entry under `unstable-book` as well?

**Comment by est31** [comment]
> does this need an entry under unstable-book as well?

The requirement for an entry has been removed. You can add one (with meaningful content) if you want to, if not, a stub will be generated automatically.

File: src/test/compile-fail/feature-gate-serial_tests.rs
Comment (bjorn3) [inline src/test/compile-fail/feature-gate-serial_tests.rs]: Please add a trailing newline.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
Lgtm when the trailing newline is added.

**Comment by shivjm** [comment]
Done, thanks!

**Review by bjorn3 — APPROVED** [review]

**Comment by brson** [comment]
Thanks for the patch.

This probably needs more discussion before landing. This is a thing that we've declined to do for many years.

**Comment by brson** [comment]
cc @nrc

**Comment by Aankhen** [comment]
Is there anything I can do to help?

**Comment by shivjm** [comment]
Is there anything I can do to help?

**Comment by nrc** [comment]
@brson do you have links to previous discussion or could you summarise here please? My opinion (uninformed by previous discussion) is that this is something I've wanted personally and I don't see why we should not support it (as long as it is feature gated etc.) being its own attribute feels kind of wrong (I'd prefer to attach it to `#[test]`, but I suppose that `#[should_panic]` is a precedent?)

**Comment by nrc** [comment]
Discussed at the dev-tools meeting today, this problem can be solved today using a mutex (I filed https://github.com/rust-lang/rust/issues/43155 to document that). Given that this solution feels somewhat ad hoc and there is workaround, we would prefer to punt on adding this in favour of custom test runners or some other long-term solution.

Thanks for the PR @shivjm and sorry to close it out after your work.

**Comment by shivjm** [comment]
Alright. Thanks for letting me know.

**Comment by yshui** [comment]
@nrc I don't see why this should be implemented in the test framework. The mutex work around seems to have some catches, and is non-trivial to get right.

**Comment by DevQps** [comment]
For everyone getting to this point: This solution is also implemented in a separate crate now: [`serial_test`](https://crates.io/crates/serial_test)

**Pull request closed without merging**
